### ISSUE TITLE
Suppling pot file for easier translation

### ISSUE THREAD

**Issue by scootergrisen** [vim/vim#2568]
I am thinking about translating Vim to danish language.
As a translator it is easier for me if i can just have a pot file to start my translation from and then from time to time i can get a updated pot file when i want to update the translation po file.

That way i don't have do download/build/install all kinds of things that i really don't need when all i want to do is translate.

So can we get a **vim/src/po/vim.pot** ?

**Comment by chrisbra** [comment]
you can do this using the provided po files. Make sure to read `https://github.com/vim/vim/blob/master/src/po/README.txt` it should explain everything to get started.

**Comment by scootergrisen** [comment]
Du you mean downloading the source and running make?
That is what im trying to avoid having to do.

**Comment by k-takata** [comment]
I understand your request, and I know some other projects provide `*.pot` for making translation easier.
However, it might require additional work to Bram.
Additionally, sometimes translators might need to compile Vim by themselves so that understand the context of messages for better translation.

**Comment by solbu** [comment]
I support this. I translate for norwegian, and unless upstream provide an updated pot file, I don't translate. and I know others do the same.

**Comment by scootergrisen** [comment]
What needs to be done so we can have the pot file?

**Comment by chrisbra** [comment]
> What needs to be done so we can have the pot file?

I don't know. Anybody knows what needs to be done to get .pot files. How are other projects handling that?

**Comment by solbu** [comment]
What other projects do is to ship the potfile with the code.
Some exclude it from revision control, which I can partly understand. But don't remove it from the source tarball.

What should be done here is that the pot file should be included with the distributed source code.

**Comment by chrisbra** [comment]
Okay let me rephrase: How is the pot file then generated/created in the first place? And how does it work together with the po file we already have. Are they somehow generated?

**Comment by scootergrisen** [comment]
Do you mean this?:

    make vim.pot
    msginit -l xx

https://github.com/vim/vim/blob/master/src/po/README.txt#L55-L56

**Comment by tonymec** [comment]
@chrisbra : Have a look at https://en.wikipedia.org/wiki/Gettext — from my nonspecialist's POV it looks like it could put you on the right way.

Best regards,
Tony.

**Comment by tonymec** [comment]
@solbu : With the master source on a github repository with a Mercurial mirror, anything excluded from revision control won't be got by anyone cloning and pulling from either of these repos.

**Comment by solbu** [comment]
My point exactly. Upstream should _NOT_ exclude the potfile from distribution.

**Comment by chrisbra** [comment]
> make vim.pot
> msginit -l xx

ah thanks, that wasn't clear to me. So basically the pot file is generated automatically. That leaves the question, why should we check in generated files at all into the repository? I thought it is general consensus to not commit auto generated files inside your vcs. So why can't you run that make command whenever you really need that pot file? Or do you only want to have it in the source tarball and we are not talking about distributing it with the source code?

**Comment by k-takata** [comment]
How about generating `vim.pot` when running `make`?

```diff
--- a/src/Makefile
+++ b/src/Makefile
@@ -1967,6 +1967,7 @@ xxd/xxd$(EXEEXT): xxd/xxd.c
 
 # Build the language specific files if they were unpacked.
 # Generate the converted .mo files separately, it's no problem if this fails.
+# *.pot will be also generated for the convenience of translators.
 languages:
 	@if test -n "$(MAKEMO)" -a -f $(PODIR)/Makefile; then \
 		cd $(PODIR); \
@@ -1976,6 +1977,10 @@ languages:
 		cd $(PODIR); \
 		  CC="$(CC)" $(MAKE) prefix=$(DESTDIR)$(prefix) converted; \
 	fi
+	-@if test -n "$(MAKEMO)" -a -f $(PODIR)/Makefile; then \
+		cd $(PODIR); \
+		  $(MAKE) prefix=$(DESTDIR)$(prefix) $(VIMNAME).pot; \
+	fi
 
 # Update the *.po files for changes in the sources.  Only run manually.
 update-po:
```

To include the pot file in the distribution, we might need to update `Filelist`.

```diff
--- a/Filelist
+++ b/Filelist
@@ -919,6 +919,7 @@ LANG_SRC = \
 		src/po/Make_mvc.mak \
 		src/po/sjiscorr.c \
 		src/po/*.po \
+		src/po/vim.pot \
 
 # the language files for the Win32 lang archive
 LANG_DOS = \
```

I don't know how Bram creates the source package, though.

**Comment by tonymec** [comment]
IIUC every .po file should depend on the .pot file and each .mo file depends on the corresponding .po file. This would (IIUC) mean recompiling all translations whenever I run `make || make install && echo 'exit status' $?` for a vim with `+multi_lang`. Or is the .pot → .po conversion strictly a (human) translator's job?

Best regards,
Tony.

**Comment by k-takata** [comment]
> Or is the .pot → .po conversion strictly a (human) translator's job?

Yes, it's translator's job.

**Comment by solbu** [comment]
> Yes, it's translator's job.

In vim, that is a translators job, yes. But I don't know any other project that require every user to independently recreate the pot-file and use that to keep the po file updated.

Normally, updating the pot and po-file with the new strings in almost every other project _is the maintainers job_, not the translators.

On this issue I disagree with the vim maintainers, but that is ok. :-)

**Comment by chrisbra** [comment]
Sorry, you previously said:

> What other projects do is to ship the potfile with the code. Some exclude it from revision control, which I can partly understand

So for those projects you would also have to re-create the pot-file and use that to keep the po file updated, right?

I mean, this is a simple matter of one single command you need to issue. Is this such a big issue to run `make vim.pot` and start translating? You can then recreate the `.po` file from it and send a PR with that diff. That does not sound too bad after all. Plus auto generated files really shouldn't really go into a vcs.

**Comment by solbu** [comment]
> So for those projects you would also have to re-create the pot-file and use that to keep the po file updated, right?

No. Those that exclude the pot file from revision control still ship the pot file in the tarball release.
So those who want to translate need to fetch the latest tarbal and use the included pot file.
Projects that exclusively use github tags as release points, include the potfile in the github revision control.

**Comment by solbu** [comment]
> I mean, this is a simple matter of one single command you need to issue. Is this such a big issue to run make vim.pot and start translating? 

For some translators that is a burden to big for them. Usually, for configure to run successfully you need to install the compile dependensies. A lot of translators are not programmers and don't know how to figure out what to install when configure complains that foo.h is missing.

>Plus auto generated files really shouldn't really go into a vcs.

Then why are there any po files in this repo? :-)
If xx.po doesn't exist, running «make xx.po» generates a file, does it not? :-)

Most generated files should be excluded from revision control, yes, but not all. The pot file is one such example.

**Comment by chrisbra** [comment]
Okay, but that works only, if those projects do not use the "Download source zip/tar archive link" from Github.

Let me conclude: We don't need the pot file in the source repository but might include the pot file in the source archive available from ftp://ftp.vim.org/pub/vim/pc/ and ftp://ftp.vim.org/pub/vim/unix/

I'll leave this open for Bram to decide if he wants to change his script to also include the pot file before uploading it to the ftp server. 

@brammool I also just noticed, that the "All files" link on the vim.org page  https://www.vim.org/sources.php still refers to the vim7.4 archive.

thanks.

**Comment by chrisbra** [comment]
> Usually, for configure to run successfully you need to install the compile dependensies. A lot of translators are not programmers and don't know how to figure out what to install when configure complains that foo.h is missing.

I consider us one of the projects, where ./configure is not needed:
```
0 116905 chrisbra@debian ~/code/vim-src/src/po (git)-[master]- % ( cd ../.. && make distclean >/dev/null ); rm -f vim.pot; make vim.pot
cd ..; OLD_PO_FILE_INPUT=yes OLD_PO_FILE_OUTPUT=yes xgettext --default-domain=vim \
        --add-comments --keyword=_ --keyword=N_ --keyword=NGETTEXT:1,2 \
        *.c if_perl.xs GvimExt/gvimext.cpp globals.h if_py_both.h vim.h
xgettext: warning: file 'if_perl.xs' extension 'xs' is unknown; will try C
mv -f ../vim.po vim.pot
```
no need to compile the binary or run configure.

**Comment by chrisbra** [comment]
Oh and regarding:
> Then why are there any po files in this repo? :-)
> If xx.po doesn't exist, running «make xx.po» generates a file, does it not? :-)

Well, we need to store the translations somewhere, right? If `make xx.po` already translates correctly, we can drop those files. I don't think this is what you want?

**Comment by solbu** [comment]
> Well, we need to store the translations somewhere, right?

I was simply refering to the statement that no generated files should be in revision control, and those is an exampel of generated files. Granted, generated by the translators, but generated none the less. :-)

**Comment by k-takata** [comment]
> Those that exclude the pot file from revision control still ship the pot file in the tarball release.

My above patch is for this purpose. Not sure this is enough for Bram, though.

**Comment by tonymec** [comment]
@solbu : my understanding was that Christian meant "files that make can generate" or "files automatically generated". If you include human authors, then by definition all files are generated (possibly by some human) so they should not be included, since the user could rewrite them from scratch. Is that what you want? Or do you mean that if a .po file is missing, make should phone the appropriate translator, chain him to the keyboard all other business ceasing (waking him up if necessary) and hold him there until the .po file has been written?

Just kidding.

**Comment by solbu** [comment]
I think we are getting side tracked, so I'm ending this. :-)
On issues I have strong opinions on I _can_ be quite pedantic and subborn, in order to make a point. :-)

**Comment by tonymec** [comment]
:-) So can I, at times. Usually I *try* to let whoever I'm speaking to know that I'm speaking tongue-in-cheek, but I don't always succeed.

**Comment by brammool** [comment]
The instructions for generating the .pot file are simple and straightforward.  If a translator has problems with that he can ask for help.  The translator will need to run Vim anyway to figure out how messages are used. and also needs to runs Vim with the translations to check them in context.  In fact, I would doubt any non-Vim user would be a good translator.  The meaning of terms would be too unclear.  Ideally a translator would use Vim with the translations regularly to check for any mistakes.  And fetch a new version once in a while to check for new or changed messages (but I know this doesn't happen much).

Anyway, providing a .pot file is not going to help anyone but translators that don't know Vim.

**Comment by brammool** [comment]
> @brammool I also just noticed, that the "All files" link on the vim.org page  https://www.vim.org/sources.php still refers to the vim7.4 archive.

Thanks for noticing, I'll update to 8.1.

-- 
ARTHUR: (as the MAN next to him is squashed by a sheep) Knights!  Run away!
   Midst echoing shouts of "run away" the KNIGHTS retreat to cover with the odd
   cow or goose hitting them still.  The KNIGHTS crouch down under cover.
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by scootergrisen** [comment]
Why not try having a pot file for a while then see the feedback.

**Comment by scootergrisen** [comment]
I translate a lot of different software and i find that the more difficult is it so update the translation the less likely i am to update the translation and i keep waiting until later and later and later.

Like now. I wanted to update the vim po file so i checked to see if there was a pot but no so i dont update it.
Had it been there i would have started updating the translation.

I dont think that the translator necessarily need to be using the software they translate regularly or even be on the same platform or test ever string in the software.
If i am in doubt about a string i can check the source code and from there i might be able to see how it should be translated.

If you think it is easy to generate the pot file then it is properly because you have it all set up.
Imagine you dont have it set up all the steps and time you have to do just in order to generate the pot file and then you have to keep this setup for next time you want to generate the pot file or do it all again.

And then you can multiply this by the numbers of translators/language.
So if the software is translated into 40 language then 40 people will have to go this.
Instead of just doing it one time and then all 40 people can use that.

If i can have a updated pot file once in a while it is just so much easier to update the translation.
Then i only have to files. The po file i have and the pot file which i can download when i need to update the translation.
Compare that to having to install a Linux distribution in VirtualBox and download git and all kinds of software to be able to run and figuring out the package names and how to get the pot file from the guest system to the host system.

And that is just for the po file.
Then the translator also have to figure out how to update the menus .vim files and good luck figuring that out.
And the .nsi file.
And other files.
So multiple formats also dont make updating the software easy.

Even if you dont want it included the pot in the source it would be nice to have generated on the side.
Like it would be on a page about translation and could be generated/updated some time because each release so translators have some time to update the translastions.

Another option is to update all the po files automatically so that translators dont have to.

**Comment by brammool** [comment]
A .pot file is not sufficient to do the translation.  The steps are detailed in https://github.com/vim/vim/blob/master/src/po/README.txt#L63
It's a little bit of work, but not complicated.
As a bonus, this will run a check script to make sure the translations are not broken.
It does require understanding of the program, Vim in this case, to do proper translations.
Doing that just from a pot file is very likely to result in a bad translation.

**Comment by scootergrisen** [comment]
I dont think it should be up to each translator to remove empty strings and stuff like that because it might not work on Solaris.
Can stuff like that not be done by the code?
It would be better if the translators can just focus on translating the strings and not having to worry about empty and obsolete strings and stuff like that.

I understand you want good translations by people that use vim but i dont think a translation will be bad just because you dont use vim each time you update the translations.
It seems like translators are almost forces to use vim to translate vim.
I dont think that is good.
I think translators should be able to use whatever software they like.
Imagine i translate 10 different text editors and then i have to translate vim in vim and notepad++ in notepad++ etc.
